pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,346 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains transforms and helpers functions for decomposing arbitrary two-qubit
+unitary operations into elementary gates.
+"""
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane import math
+
+from .single_qubit_unitary import zyz_decomposition
+
+# This gate E is called the "magic basis". It can be used to convert between
+# SO(4) and SU(2) x SU(2). For A in SO(4), E A E^\dag is in SU(2) x SU(2).
+E = qml.math.array([[1, 1j, 0, 0], [0, 0, 1j, 1], [0, 0, 1j, -1], [1, -1j, 0, 0]]) / np.sqrt(2)
+Et = qml.math.T(E)
+Edag = qml.math.conj(Et)
+
+# Helpful to have static copies of these since they are needed in a few places.
+CNOT01 = qml.CNOT(wires=[0, 1]).matrix
+CNOT10 = qml.math.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]])
+SWAP = qml.SWAP(wires=[0, 1]).matrix
+
+
+def _perm_matrix_from_sequence(seq):
+    """Construct a permutation matrix based on a provided permutation of integers.
+
+    This is used in the two-qubit unitary decomposition to permute a set of
+    simultaneous eigenvalues/vectors into the same order.
+    """
+    mat = qml.math.zeros((4, 4))
+
+    for row_idx in range(4):
+        mat[row_idx, seq[row_idx]] = 1
+
+    return mat
+
+
+def _convert_to_su4(U):
+    r"""Check unitarity of a 4x4 matrix and convert it to :math:`SU(4)` if possible.
```

Comment:
From my reading of the code, sounds like this should be?
```suggestion
    r"""Check unitarity of a 4x4 matrix and convert it to :math:`SU(4)` if the determinant is not 1.
```